rfc/text/advanced/ap_35_authentication_cryptosign.md

```diff
@@ -1,3 +1,937 @@
+2. the router sends the signature as part of its challenge to the client, in `CHALLENGE.extra.signature|string`
+
+Again, in this case, the *Router* includes a trustroot and certificate for the client to verify.
+
+
+#### Computing the Signature
+
+The challenge sent by the router is a 32 bytes random value, encoded as a Hex string in `CHALLENGE.extra.challenge|string`.
+
+When no channel binding is active, the Ed25519 signature over the 32 bytes message MUST be computed using the WAMP-Cryptosign *private key* of the authenticating client.
+
+When channel binding is active, the challenge MUST first be XOR'ed bytewise with the channel ID, e.g. the 32 bytes from TLS with channel binding `"tls-unique"`, and the resulting message (which again has length 32 bytes) MUST be signed using the WAMP-Cryptosign *private key* of the authenticating client.
+
+The client MUST return the concatenation of the signature and the message signed (96 bytes) in the `AUTHENTICATE` message.
+
+
+#### Example Message Flow
+
+A typical authentication begins with the client sending a `HELLO` message specifying the `cryptosign` method as (one of) the authentication methods:
+
+{align="left"}
+```javascript
+[1, "realm1", {
+    "roles": {/* see below */},
+    "authmethods": ["cryptosign"],
+    "authid": "client01@example.com",
+    "authextra": {
+        "pubkey": "545efb0a2192db8d43f118e9bf9aee081466e1ef36c708b96ee6f62dddad9122",
+        "channel_binding": null
+    }
+}]
+```
+
+The `HELLO.Details.authmethods|list` is used by the client to announce the authentication methods it is prepared to perform. For WAMP-Cryptosign, this MUST include `"cryptosign"`.
+
+The `HELLO.Details.authid|string` is the authentication ID (e.g. username) the client wishes to authenticate as. For WAMP-Cryptosign authentication, this MAY be provided. If no `authid` is provided, the router SHOULD automatically chose and assign an `authid` (e.g. the Hex encode public key).
+
+The `HELLO.Details.authextra|dict` contains the following members for WAMP-Cryptosign:
+
+| Field             | Type     | Required | Description                                                                          |
+|-------------------|----------|----------|--------------------------------------------------------------------------------------|
+| pubkey            | string   | yes      | The client public key (32 bytes) as a Hex encoded string, e.g. `545efb0a2192db8d43f118e9bf9aee081466e1ef36c708b96ee6f62dddad9122` |
+| channel_binding   | string   | no       | If TLS channel binding is in use, the TLS channel binding type, e.g. `"tls-unique"`. |
+| challenge         | string   | no       | A client chosen, random challenge (32 bytes) as a Hex encoded string, to be signed by the router. |
+| trustroot         | string   | no       | When the client includes a client certificate (see below), the Ethereum address of the trustroot of the certificate, e.g. `0x72b3486d38E9f49215b487CeAaDF27D6acf22115` |
```

Comment:
I think we should add more words about trustroots. Maybe describe possible cases, like the most common one is to rely on underlying (browsers, OS) root certificates. But this option allows users to choose and specify independent blockchain-based authority for trusted certificates. And maybe also say, that Ethereum is not the one possible option, that this can be extended in the future.
Because i think some users may be afraid that they need blockchain for wamp to work %) Which is of course not true :)

Reply:
I will add a background section, and the two options: A) having an single operator controlled trustroot (a root CA self-signed cert), and B) having a shared trustroot that is stored on-chain

Reply:
Done